### PR TITLE
Core/Scripts: Boss Jedoga Shadowseeker rewrite

### LINKED ISSUE #5182: [Instances] Ahn'Kahet - Jedoga Shadowseeker
Boss id - http://www.wowhead.com/npc=29310
Issues:
0. The trigger that stays near the water infront of boss or in other words NPC_JEDOGA_CONTROLLER = 30181 is visible to regular players, when it shouldn't.
I. After clearing the 15 Twilight Initiates http://www.wowhead.com/npc=30114 boss should descend and activate /that part works/, the problem is after it 15 Twilight Volunteers should spawn http://www.wowhead.com/npc=30385 in not active state just kneeling and waiting for her to call them for sacrifice mentioned on point II. The problem is in the script there is only mentioned the ID of the  Initiates and it use them instead of the Volunteers /in simple words it is scripted right, but wrong npc is being used/. So after boss gets activated 15 Initiates should be replaced by 15 Volunteers.
II. Once that is done the sacrifice itself use Initiates so the id and npc should be changed to Volunteer /not sure if there must be some npc state changes, when I tried it I got it not atackable just running./
III. Since 3.3.2 this encounter has been nerfed, still Trinity's script is staying at boss old state. What should be changed - boss should only 1 time fly up above the ground at 50% of her hp /and this is the change/ in not targetable state and use http://www.wowhead.com/spell=56150 on Twilight Volunteer to get http://www.wowhead.com/spell=56219, if the party doesn't kill the npc before that. Atm she use the old boss script and do it many times.
IV. Here I will mention the achievement part for http://www.wowhead.com/achievement=2056. The script is working 50% atm, the conditions are set /though id of initate should be changed with volunteer/ and atm the thing that is bugged is that it caunt boss sacrifice as failing the achievement, there should be set extra condition that if volunteer get sacrificed http://www.wowhead.com/spell=56150 -> achievement return true.
Connected with issue - https://github.com/TrinityCore/TrinityCore/issues/3276, just mentioning the ach part to have it here 2.
V. In case of wipe the 15 Initiates that trigger boss activate should respawn-reset, currently only 2 do.

TDB: 11.45
Win 7
a3e739e7261da501d1b901c2c738a7231b570eae
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/65299-instances-ahn-kahet-jedoga-shadowseeker?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sirikfoll**
**Changes proposed:**
-  Added all missing spells and visuals
-  Added Twilight volunteers in the fight
-  Fixed Jedoga Controller's behavior, mostly moved to boss_prince_taldaram, where they belong
-  Updated script register model
-  Misc updates and improvements
-  Updated instance model

**Target branch(es):** 3.3.5/master
- [X] 3.3.5
- [X] master

**Issues addressed:** Closes #5182

**Tests performed:** Built and tested, alone.

### CODE REVIEW

File: src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp [hunk 1]
```diff
@@ -15,589 +15,487 @@
...
 
-        uint32 uiOpFerTimer;
-        uint32 uiCycloneTimer;
-        uint32 uiBoltTimer;
-        uint32 uiThunderTimer;
+typedef std::pair<Position, Position> VolunteerPositionPair;
+std::vector<VolunteerPositionPair> VolunteerSpotPositions =
+{
+    { { 400.7701f, -784.8928f, -31.60143f }, { 365.9514f, -719.1235f, -16.17974f } },
+    { { 397.3595f, -788.5157f, -31.59679f }, { 359.7433f, -715.017f,  -16.17974f } },
```
Comment (sirikfoll) [inline src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp]: Can I keep this part as it is or should I be using something like bellow?
```cpp
std::make_pair(Position(400.7701f, -784.8928f, -31.60143f), Position(365.9514f, -719.1235f, -16.17974f)),

File: src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp [hunk 2]
```diff
@@ -15,589 +15,487 @@
...
 
-                    if (killer->GetTypeId() == TYPEID_PLAYER)
-                        boss->AI()->DoAction(ACTION_INITIAND_KILLED);
+                    _selectedVolunteerGUID = Trinity::Containers::SelectRandomContainerElement(_volunteerGUIDS);
+                    if (Creature* volunteer = ObjectAccessor::GetCreature(*me, _selectedVolunteerGUID))
+                        volunteer->AI()->DoAction(ACTION_CHOSEN);
+                    break;
+                case EVENT_SUMMON_VOLUNTEER:
+                {
+                    uint32 pos = std::distance(_volunteerGUIDS.begin(), std::find(_volunteerGUIDS.begin(), _volunteerGUIDS.end(), _selectedVolunteerGUID));
```
Comment (sirikfoll) [inline src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp]: Is there any problem using this?

File: src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp (hunk 1 above)
Comment (ccrs) [reply src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp]: no need to

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 3]
```diff
@@ -0,0 +1,62 @@
+UPDATE `creature_template` SET `unit_flags`=33587264, `InhabitType`=4 WHERE `entry` IN (29310,31465); -- Jedoga
+UPDATE `creature` SET `MovementType`=0, `spawndist`=0 WHERE `guid`=131953;
+UPDATE `creature_addon` SET `auras`='56075 56327' WHERE `guid`=131953; -- Jedoga
+UPDATE `creature_template` SET `ScriptName`='npc_twilight_volunteer' WHERE `entry`=30385;
+UPDATE `creature_template` SET `ScriptName`='' WHERE `entry` IN(30114,30181);
+DELETE FROM `creature` WHERE `id` IN(30114,30181);
+DELETE FROM `creature_addon` WHERE `guid` IN (132063,132066,132067,132068,132069,132070,132071,132072,132065);
+DELETE FROM `linked_respawn` WHERE `guid` IN (131997,131998,131999,132000,132001,132002,132003,132004,132005,132006,132007,132008,132009,132010,132011);
```
Comment (ccrs) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: linked_respawn doesnt exist on head right?

File: src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp [hunk 4]
```diff
@@ -15,589 +15,487 @@
...
+    { 420.1999f, -727.0132f, 28.88036f, 2.042035f },
+    { 375.4977f, -707.3635f, -16.0964f, 2.426008f }
+};
 
-        uint32 uiOpFerTimer;
-        uint32 uiCycloneTimer;
-        uint32 uiBoltTimer;
-        uint32 uiThunderTimer;
+typedef std::pair<Position, Position> VolunteerPositionPair;
+std::vector<VolunteerPositionPair> VolunteerSpotPositions =
```
Comment (ccrs) [inline src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp]: const

File: src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp [hunk 5]
```diff
@@ -15,589 +15,487 @@
...
+        _volunteerGUIDS.clear();
+        _selectedVolunteerGUID.Clear();
+        me->SummonCreatureGroup(SUMMON_GROUP_INITIATES);
+
+        if (TempSummon* controller = me->SummonCreature(NPC_JEDOGA_CONTROLLER, JedogaControllerPositions[0], TEMPSUMMON_MANUAL_DESPAWN))
+            controller->CastSpell(me, SPELL_BEAM_VISUAL_JEDOGA);
+        if (TempSummon* controller = me->SummonCreature(NPC_JEDOGA_CONTROLLER, JedogaControllerPositions[1], TEMPSUMMON_MANUAL_DESPAWN))
+            controller->CastSpell(me, SPELL_BEAM_VISUAL_JEDOGA);
+
+        events.ScheduleEvent(EVENT_INTRO_SAY, Minutes(2), 0, PHASE_INTRO);
```
Comment (ccrs) [inline src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp]: I think this is triggered by LoS, I remember confirming it on pre cata rework

File: src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp (hunk 2 above)
Comment (ccrs) [reply src/server/scripts/Northrend/AzjolNerub/Ahnkahet/boss_jedoga_shadowseeker.cpp]: not really, if it works as intended

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql (hunk 3 above)
Comment (sirikfoll) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Still exists, will trigger a DB error if the values are kept there

**Comment by Keader** [comment]
Click on Merge button 👇

**Comment by hiMalleus** [comment]
I tested this and it has a few issues:

- The summoned creature's pathing is not working properly since they don't come through the stairs. Instead they go through terrain.

- The Sacrifice event only happens once.

**Comment by sirikfoll** [comment]
@hiMalleus  pathing issue is something with waypoints system

Sacrifice should only happen once, that's the correct behavior.

**Comment by Keader** [comment]
yay !

**Pull request merged**